File: acceptance_test.py
#!/usr/bin/env python3
"""Acceptance test for xmlgate, written before the product ships (Article 9.1).

The test is the promise. If any of these fail, the thing does not ship:

  1. The clean control passes through byte-identical. A repair tool that rewrites a
     healthy file is a tool an engraver cannot trust with a finished score.
  2. Each repairable class is repaired, and the repaired file re-checks CLEAN.
     A fix that leaves the file still broken is not a fix.
  3. Repair is lossless where it claims to be: the MARKUP repair changes nothing
     but the closing bracket, and the UNSCALED repair changes nothing but durations.
  4. The unknowable case is REFUSED and the file is left untouched. This is the one
     the product is actually sold on - guessing here corrupts a score silently.
  5. Refusal names the location. "Something is wrong" is not a deliverable.

Run: python acceptance_test.py    (exit 0 = all pass)
"""
import os
import shutil
import sys
import tempfile
import xml.etree.ElementTree as ET

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import xmlgate

HERE = os.path.dirname(os.path.abspath(__file__))
TESTS = os.path.join(HERE, 'tests')
CASES = []


def case(name):
    def deco(fn):
        CASES.append((name, fn))
        return fn
    return deco


def run(fixture):
    """Check a fixture, write any repair to a temp file, return (verdicts, out_path, changed)."""
    return run_path(os.path.join(TESTS, fixture), os.path.basename(fixture))


def run_path(src, basename=None):
    tree, findings, changed = xmlgate.check(src)
    out = os.path.join(tempfile.mkdtemp(), basename or os.path.basename(src))
    if changed:
        tree.write(out, encoding='UTF-8', xml_declaration=True)
    else:
        shutil.copyfile(src, out)
    return findings, out, changed


def verdicts(findings):
    return [f[2] for f in findings]


def durations(path):
    return [int(d.text) for d in ET.parse(path).getroot().iter('duration')]


def tuplet_marks(path):
    return [(t.get('type'), t.get('number')) for t in ET.parse(path).getroot().iter('tuplet')]


@case('1. clean control passes through byte-identical')
def _clean():
    findings, out, changed = run('a-clean.musicxml')
    assert verdicts(findings) == ['CLEAN'], verdicts(findings)
    assert changed is False, 'the tool modified a healthy file'
    a = open(os.path.join(TESTS, 'a-clean.musicxml'), 'rb').read()
    b = open(out, 'rb').read()
    assert a == b, 'clean file came back changed (%d vs %d bytes)' % (len(a), len(b))


@case('2a. open tuplet bracket is classified MARKUP, repaired, and re-checks clean')
def _markup():
    findings, out, changed = run('b-markup.musicxml')
    assert verdicts(findings) == ['MARKUP'], verdicts(findings)
    assert changed is True, 'nothing was repaired'
    again, _, _ = run_path(out)
    assert verdicts(again) == ['CLEAN'], 'repaired file does not re-check clean: %s' % verdicts(again)


@case('2b. unscaled durations are classified UNSCALED, repaired, and re-check clean')
def _unscaled():
    findings, out, changed = run('c-unscaled.musicxml')
    assert verdicts(findings) == ['UNSCALED'], verdicts(findings)
    assert changed is True, 'nothing was repaired'
    again, _, _ = run_path(out)
    assert verdicts(again) == ['CLEAN'], 'repaired file does not re-check clean: %s' % verdicts(again)


@case('3a. MARKUP repair touches the bracket and nothing else')
def _markup_lossless():
    src = os.path.join(TESTS, 'b-markup.musicxml')
    _, out, _ = run('b-markup.musicxml')
    assert durations(src) == durations(out), 'durations were altered by a markup repair'
    assert tuplet_marks(src) == [('start', '1')], tuplet_marks(src)
    assert tuplet_marks(out) == [('start', '1'), ('stop', '1')], tuplet_marks(out)


@case('3b. UNSCALED repair touches durations and nothing else')
def _unscaled_lossless():
    src = os.path.join(TESTS, 'c-unscaled.musicxml')
    _, out, _ = run('c-unscaled.musicxml')
    assert durations(src) == [12, 12, 12, 24, 24, 24], durations(src)
    assert durations(out) == [8, 8, 8, 24, 24, 24], durations(out)
    assert tuplet_marks(src) == tuplet_marks(out), 'a duration repair moved a bracket'


@case('4. unknowable damage is REFUSED and the file is left untouched')
def _refused():
    findings, out, changed = run('d-unknowable.musicxml')
    assert verdicts(findings) == ['REFUSED'], verdicts(findings)
    assert changed is False, 'the tool guessed at damage it cannot know'
    a = open(os.path.join(TESTS, 'd-unknowable.musicxml'), 'rb').read()
    assert a == open(out, 'rb').read(), 'a refused file was still rewritten'


@case('5. refusal names part, measure and the offending durations')
def _refusal_names_it():
    findings, _, _ = run('d-unknowable.musicxml')
    pid, num, verdict, detail = findings[0]
    assert pid == 'P1' and num == '1', (pid, num)
    assert '[8, 9, 8]' in detail, 'the refusal does not show the durations: %s' % detail
    assert '25' in detail and '24' in detail, 'the refusal does not show the sums: %s' % detail


@case('6. no false alarm on any file of the third-party corpus')
def _no_false_alarms():
    import glob
    files = sorted(glob.glob(os.path.join(HERE, 'corpus', '*.musicxml')))
    assert len(files) >= 60, 'corpus is missing; see corpus/PROVENANCE.md (%d files)' % len(files)
    alarms = []
    for f in files:
        try:
            _, findings, changed = xmlgate.check(f)
        except Exception as e:
            alarms.append('%s crashed: %s: %s' % (os.path.basename(f), type(e).__name__, e))
            continue
        assert changed is False, '%s: a valid file was modified' % os.path.basename(f)
        for pid, num, v, d in findings:
            if v in (xmlgate.REFUSED, xmlgate.MARKUP, xmlgate.UNSCALED):
                alarms.append('%s part %s m%s: %s' % (os.path.basename(f), pid, num, v))
    assert not alarms, 'valid files raised %d alarm(s):\n          %s' % (
        len(alarms), '\n          '.join(alarms))


@case('7. nested tuplets are declined by name, not judged')
def _nested_declined():
    src = os.path.join(HERE, 'corpus', '23d-Tuplets-Nested.musicxml')
    if not os.path.exists(src):
        raise AssertionError('corpus is missing; see corpus/PROVENANCE.md')
    _, findings, changed = xmlgate.check(src)
    assert changed is False, 'a nested-tuplet file was modified'
    assert findings and all(v == xmlgate.UNCHECKABLE for _, _, v, _ in findings), \
        [f[2] for f in findings]
    assert all('nested' in d for _, _, _, d in findings), \
        'the reason given is not nesting: %s' % [f[3][:40] for f in findings]


def main():
    print('xmlgate acceptance test')
    print('-' * 74)
    failed = 0
    for name, fn in CASES:
        try:
            fn()
            print('  PASS  %s' % name)
        except AssertionError as e:
            failed += 1
            print('  FAIL  %s\n          %s' % (name, e))
        except Exception as e:
            failed += 1
            print('  ERROR %s\n          %s: %s' % (name, type(e).__name__, e))
    print('-' * 74)
    print('%d of %d passed' % (len(CASES) - failed, len(CASES)))
    return 1 if failed else 0


if __name__ == '__main__':
    sys.exit(main())

File: web/acceptance_test.js
/* Acceptance test for the browser port. The same nine promises as
 * ../acceptance_test.py, driven from the same four fixtures and the same 22-file
 * third-party corpus, plus two the Python version could not make:
 *
 *   10. every repair is a surgical edit — no byte outside an edited span moves.
 *   11. the port agrees with the Python original, verdict for verdict, on all 64 files.
 *
 * Case 11 reads verdicts.json, written by ../compare_ports.py in the same run.
 * Run: node acceptance_test.js   (exit 0 = all pass)
 */
'use strict';
const fs = require('fs');
const path = require('path');
const xmlgate = require('./xmlgate.js');

const HERE = __dirname;
const TESTS = path.join(HERE, '..', 'tests');
const CORPUS = path.join(HERE, '..', 'corpus');
const CASES = [];
const case_ = (name, fn) => CASES.push([name, fn]);

function assert(cond, msg) { if (!cond) throw new Error(msg); }
function read(p) { return fs.readFileSync(p, 'utf8'); }

function run(fixture) {
  const src = read(path.join(TESTS, fixture));
  const r = xmlgate.check(src);
  return { src, out: xmlgate.repair(src, r.edits), ...r };
}
const verdicts = (f) => f.map((x) => x.verdict);

function durations(text) {
  return xmlgate._internal.iter(xmlgate.parse(text), 'duration')
    .map((el) => parseInt(xmlgate._internal.textOf(el, text), 10));
}
function tupletMarks(text) {
  return xmlgate._internal.iter(xmlgate.parse(text), 'tuplet')
    .map((el) => [el.attrs.type, el.attrs.number || '1'].join(':'));
}

/* Reverse every edit against the repaired text. If the original comes back
 * byte for byte, then nothing outside the edited spans was touched. */
function unchangedOutsideEdits(src, out, edits) {
  const asc = edits.slice().sort((a, b) => a.start - b.start);
  let rebuilt = '', cursorOut = 0, cursorSrc = 0;
  for (const e of asc) {
    const lead = e.start - cursorSrc;
    rebuilt += out.slice(cursorOut, cursorOut + lead);
    cursorOut += lead + e.text.length;
    rebuilt += src.slice(e.start, e.end);
    cursorSrc = e.end;
  }
  rebuilt += out.slice(cursorOut);
  return rebuilt === src;
}

case_('1. clean control passes through byte-identical', () => {
  const r = run('a-clean.musicxml');
  assert(JSON.stringify(verdicts(r.findings)) === '["CLEAN"]', verdicts(r.findings));
  assert(r.changed === false, 'the tool modified a healthy file');
  assert(r.out === r.src, 'clean file came back changed');
});

case_('2a. open tuplet bracket is classified MARKUP, repaired, and re-checks clean', () => {
  const r = run('b-markup.musicxml');
  assert(JSON.stringify(verdicts(r.findings)) === '["MARKUP"]', verdicts(r.findings));
  assert(r.changed === true, 'nothing was repaired');
  const again = xmlgate.check(r.out);
  assert(JSON.stringify(verdicts(again.findings)) === '["CLEAN"]',
    'repaired file does not re-check clean: ' + verdicts(again.findings));
});

case_('2b. unscaled durations are classified UNSCALED, repaired, and re-check clean', () => {
  const r = run('c-unscaled.musicxml');
  assert(JSON.stringify(verdicts(r.findings)) === '["UNSCALED"]', verdicts(r.findings));
  assert(r.changed === true, 'nothing was repaired');
  const again = xmlgate.check(r.out);
  assert(JSON.stringify(verdicts(again.findings)) === '["CLEAN"]',
    'repaired file does not re-check clean: ' + verdicts(again.findings));
});

case_('3a. MARKUP repair touches the bracket and nothing else', () => {
  const r = run('b-markup.musicxml');
  assert(JSON.stringify(durations(r.src)) === JSON.stringify(durations(r.out)),
    'durations were altered by a markup repair');
  assert(JSON.stringify(tupletMarks(r.src)) === '["start:1"]', tupletMarks(r.src));
  assert(JSON.stringify(tupletMarks(r.out)) === '["start:1","stop:1"]', tupletMarks(r.out));
});

case_('3b. UNSCALED repair touches durations and nothing else', () => {
  const r = run('c-unscaled.musicxml');
  assert(JSON.stringify(durations(r.src)) === '[12,12,12,24,24,24]', durations(r.src));
  assert(JSON.stringify(durations(r.out)) === '[8,8,8,24,24,24]', durations(r.out));
  assert(JSON.stringify(tupletMarks(r.src)) === JSON.stringify(tupletMarks(r.out)),
    'a duration repair moved a bracket');
});

case_('4. unknowable damage is REFUSED and the file is left untouched', () => {
  const r = run('d-unknowable.musicxml');
  assert(JSON.stringify(verdicts(r.findings)) === '["REFUSED"]', verdicts(r.findings));
  assert(r.changed === false, 'the tool guessed at damage it cannot know');
  assert(r.out === r.src, 'a refused file was still rewritten');
});

case_('5. refusal names part, measure and the offending durations', () => {
  const r = run('d-unknowable.musicxml');
  const f = r.findings[0];
  assert(f.part === 'P1' && f.measure === '1', f.part + ' ' + f.measure);
  assert(f.detail.includes('[8, 9, 8]'), 'the refusal does not show the durations: ' + f.detail);
  assert(f.detail.includes('25') && f.detail.includes('24'),
    'the refusal does not show the sums: ' + f.detail);
});

case_('6. no false alarm on any file of the third-party corpus', () => {
  const files = fs.readdirSync(CORPUS).filter((f) => f.endsWith('.musicxml')).sort();
  assert(files.length >= 60, 'corpus is missing; see corpus/PROVENANCE.md (' + files.length + ')');
  const alarms = [];
  for (const f of files) {
    const src = read(path.join(CORPUS, f));
    let r;
    try { r = xmlgate.check(src); }
    catch (e) { alarms.push(f + ' crashed: ' + e.message); continue; }
    assert(r.changed === false, f + ': a valid file was modified');
    for (const x of r.findings) {
      if (x.verdict === 'REFUSED' || x.verdict === 'MARKUP' || x.verdict === 'UNSCALED') {
        alarms.push(f + ' part ' + x.part + ' m' + x.measure + ': ' + x.verdict);
      }
    }
  }
  assert(!alarms.length, 'valid files raised ' + alarms.length + ' alarm(s):\n          ' +
    alarms.join('\n          '));
});

case_('7. nested tuplets are declined by name, not judged', () => {
  const p = path.join(CORPUS, '23d-Tuplets-Nested.musicxml');
  assert(fs.existsSync(p), 'corpus is missing; see corpus/PROVENANCE.md');
  const r = xmlgate.check(read(p));
  assert(r.changed === false, 'a nested-tuplet file was modified');
  assert(r.findings.length && r.findings.every((x) => x.verdict === 'UNCHECKABLE'),
    JSON.stringify(verdicts(r.findings)));
  assert(r.findings.every((x) => x.detail.includes('nested')),
    'the reason given is not nesting');
});

case_('10. every repair is a surgical edit - no byte outside an edited span moves', () => {
  for (const fixture of ['b-markup.musicxml', 'c-unscaled.musicxml']) {
    const r = run(fixture);
    assert(r.edits.length > 0, fixture + ': expected edits');
    assert(unchangedOutsideEdits(r.src, r.out, r.edits),
      fixture + ': the repair moved bytes outside its own edit spans');
  }
});

case_('11. the port agrees with the Python original, verdict for verdict, on all 64 files', () => {
  const ref = path.join(HERE, 'verdicts.json');
  assert(fs.existsSync(ref), 'verdicts.json missing - run ../compare_ports.py first');
  const expected = JSON.parse(read(ref));
  const names = Object.keys(expected).sort();
  assert(names.length >= 64, 'expected at least 64 files, got ' + names.length);
  const mismatches = [];
  for (const name of names) {
    const dir = name.startsWith('tests/') ? TESTS : CORPUS;
    const src = read(path.join(dir, path.basename(name)));
    const r = xmlgate.check(src);
    const mine = r.findings.map((x) => [x.part, x.measure, x.verdict].join('|'));
    const theirs = expected[name];
    if (JSON.stringify(mine) !== JSON.stringify(theirs)) {
      mismatches.push(name + '\n            python: ' + JSON.stringify(theirs) +
                             '\n            js:     ' + JSON.stringify(mine));
    }
  }
  assert(!mismatches.length, mismatches.length + ' file(s) disagree:\n          ' +
    mismatches.join('\n          '));
});

function main() {
  console.log('xmlgate browser-port acceptance test');
  console.log('-'.repeat(74));
  let failed = 0;
  for (const [name, fn] of CASES) {
    try { fn(); console.log('  PASS  ' + name); }
    catch (e) { failed++; console.log('  FAIL  ' + name + '\n          ' + e.message); }
  }
  console.log('-'.repeat(74));
  console.log((CASES.length - failed) + ' of ' + CASES.length + ' passed');
  process.exit(failed ? 1 : 0);
}
main();
